OSOE-464: Replacing IRecipeHarvester implementations

dministro opened the pull request:
[OSOE-464](https://lombiq.atlassian.net/browse/OSOE-464)


[OSOE-464]: https://lombiq.atlassian.net/browse/OSOE-464?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

dministro committed:
Replacing IRecipeHarvester implementations

dministro committed:
Fix spelling

dministro committed:
Fixing analyzer violation

0liver commented on Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs:
```diff
@@ -140,6 +144,16 @@ private void AddFakeStore(IServiceCollection services)
     private static void AddFakeViewCompilerProvider(IServiceCollection services) =>
         services.AddSingleton<IViewCompilerProvider, FakeViewCompilerProvider>();
 
+    // We remove the existing IRecipeHarvester implementations and add a custom implementation that uses the same come
```
```suggestion
    // We remove the existing IRecipeHarvester implementations and add a custom implementation that uses the same code
```

dministro replied:
Thank you @0liver

dministro committed:
Update Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs

Co-authored-by: 0liver <0liver@users.noreply.github.com>

Piedone commented on Lombiq.Tests.UI/Services/OrchardCoreHosting/FakeRecipeHarvester.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace Lombiq.Tests.UI.Services.OrchardCoreHosting;
+
+public class FakeRecipeHarvester : IRecipeHarvester
```
Please don’t forget to create an OC issue about this, since this needs to actually be fixed there, and comment the issue link here too. Or do we need this solely for UI testing and nothing to change in OC? Also comment if there's anything special to take care of here during OC upgrades.

dministro replied:
There is an open issue for this in OC: https://github.com/OrchardCMS/OrchardCore/issues/10329

Piedone replied:
OK then!

sarahelsaig replied:
Since we already have the solution here, would you open an OCORE issue in Jira for contributing it back upstream?

dministro committed:
Adding related OC issue link to comment

sarahelsaig commented on Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs:
```diff
@@ -140,6 +144,18 @@ private void AddFakeStore(IServiceCollection services)
+    // OC issue: https://github.com/OrchardCMS/OrchardCore/issues/10329.
+    private static void AddFakeRecipeHarvester(IServiceCollection services)
+    {
+        services.RemoveAll<IRecipeHarvester>();
```
Shouldn't you specifically remove `RecipeHarvester` and `ApplicationRecipeHarvester` instead of all implementations? We don't know if others may add their own recipe harvesters, but if they do this will break. (we add our own template harvester in Lombiq.VueJs so that feels similar)

sarahelsaig commented on Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs:
```diff
@@ -140,6 +144,18 @@ private void AddFakeStore(IServiceCollection services)
+    {
+        services.RemoveAll<IRecipeHarvester>();
+
+        services.AddScoped<IRecipeHarvester, FakeApplicationRecipeHarvester>()
+            .AddScoped<IRecipeHarvester, FakeRecipeHarvester>();
```
I don't think we usually chain the service add methods like this.

sarahelsaig commented on Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs:
```diff
@@ -140,6 +144,18 @@ private void AddFakeStore(IServiceCollection services)
     private static void AddFakeViewCompilerProvider(IServiceCollection services) =>
         services.AddSingleton<IViewCompilerProvider, FakeViewCompilerProvider>();
 
+    // We remove the existing IRecipeHarvester implementations and add a custom implementation that uses the same code
+    // as OC but with a fix in RecipeHarvester.HarvestRecipesAsync to avoid sync over async issue.
```
> that uses the same code as OC

Then we shouldn't call it a fake harvester since it actually functions and fake should be reserved to stub/dummy services. How about AsyncRecipeHarvester?

dministro committed:
Renaming IRecipeHarvester implementations
Removing service method call chaining
Replacing IRecipeHarvester implementations only implemented in OrchardCore.Recipes
